File: sentil-java/src/main/java/io/github/sedislab/sentil/Monitor.java
package io.github.sedislab.sentil;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.OptionalLong;

/** A monitor for one formula. */
public final class Monitor extends NativeResource {
    Monitor(long handle) {
        super(handle, NativeLib::monitorDestroy);
    }

    /** A monitor for a formula, with the default discrete-time config. Consumes the formula. */
    public Monitor(Formula formula) throws SentilException {
        this(NativeLib.monitorCreate(formula.consume(), 0L));
    }

    /** A monitor for a formula with the given config. Consumes the formula. */
    public Monitor(Formula formula, Config config) throws SentilException {
        this(NativeLib.monitorCreate(formula.consume(), config.handle()));
    }

    /** A monitor for a formula string, with the default config. */
    public static Monitor parse(String formula) throws SentilException {
        return new Monitor(NativeLib.monitorParse(formula, 0L));
    }

    /** A monitor for a formula string with the given config. */
    public static Monitor parse(String formula, Config config) throws SentilException {
        return new Monitor(NativeLib.monitorParse(formula, config.handle()));
    }

    /** A copy of the monitored formula. */
    public Formula formula() throws SentilException {
        return new Formula(NativeLib.monitorFormula(handle()));
    }

    /** A copy of the monitor's config. */
    public Config config() throws SentilException {
        return new Config(NativeLib.monitorConfig(handle()));
    }

    /** The robustness over the trace, honoring the config's time mode. */
    public double robustness(Trace trace) throws SentilException {
        return NativeLib.monitorRobustness(handle(), trace.handle());
    }

    /** The robustness at every sample of the trace. */
    public double[] robustnessSignal(Trace trace) throws SentilException {
        return NativeLib.monitorRobustnessSignal(handle(), trace.handle());
    }

    /** The time spans where the property does not hold. */
    public List<Interval> violations(Trace trace) throws SentilException {
        return Arrays.asList(NativeLib.monitorViolations(handle(), trace.handle()));
    }

    /** The index of a variable in packed-update order, or empty if the formula skips it. */
    public OptionalLong symbolIndex(String name) throws SentilException {
        long[] index = NativeLib.monitorSymbolIndex(handle(), name);
        return index.length == 0 ? OptionalLong.empty() : OptionalLong.of(index[0]);
    }

    /** Fold one timestamped sample given as a map from variable name to value. */
    public Robustness update(double time, Map<String, Double> values) throws SentilException {
        List<String> names = new ArrayList<>(values.size());
        double[] data = new double[values.size()];
        int i = 0;
        for (Map.Entry<String, Double> entry : values.entrySet()) {
            names.add(entry.getKey());
            data[i++] = entry.getValue();
        }
        return NativeLib.monitorUpdate(handle(), time, names.toArray(new String[0]), data);
    }

    /** Fold one sample with values already in {@link #symbolIndex} order. */
    public Robustness updatePacked(double time, double[] values) throws SentilException {
        return NativeLib.monitorUpdatePacked(handle(), time, values);
    }

    /** Clear streaming state so the monitor can run a fresh trace. */
    public void reset() {
        NativeLib.monitorReset(handle());
    }
}

File: sentil-matlab/+sentil/Monitor.m
classdef Monitor < handle
    %MONITOR An offline and incremental monitor for one formula.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods
        function obj = Monitor(spec, config)
            %MONITOR Build from a formula string or a consumed sentil.Formula.
            cfg = uint64(0);
            if nargin >= 2 && ~isempty(config)
                cfg = config.Handle;
            end
            if isa(spec, 'sentil.Formula')
                obj.Handle = sentil_mex('monitor_create', spec.consume(), cfg);
            elseif ischar(spec) || isstring(spec)
                obj.Handle = sentil_mex('monitor_parse', char(spec), cfg);
            elseif isa(spec, 'uint64') && isscalar(spec)
                obj.Handle = spec;
            else
                error('sentil:monitor', 'Monitor(formulaOrString, config)');
            end
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('monitor_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function r = robustness(obj, trace)
            %ROBUSTNESS Robustness over the trace.
            obj.assertOpen();
            r = sentil_mex('monitor_robustness', obj.Handle, trace.Handle);
        end

        function r = robustness_signal(obj, trace)
            %ROBUSTNESS_SIGNAL Robustness at every sample, as a row vector.
            obj.assertOpen();
            r = sentil_mex('monitor_robustness_signal', obj.Handle, trace.Handle);
        end

        function v = violations(obj, trace)
            %VIOLATIONS The failing time spans, as [start, end] rows.
            obj.assertOpen();
            v = sentil_mex('monitor_violations', obj.Handle, trace.Handle);
        end

        function idx = symbol_index(obj, name)
            %SYMBOL_INDEX The 1-based position of a variable in update_packed order.
            obj.assertOpen();
            idx = sentil_mex('monitor_symbol_index', obj.Handle, char(name));
        end

        function r = update(obj, time, sample)
            %UPDATE Fold one timestamped sample, given as a struct or containers.Map.
            obj.assertOpen();
            [names, values] = unpack_sample(sample);
            r = sentil_mex('monitor_update', obj.Handle, double(time), names, values);
        end

        function r = update_packed(obj, time, values)
            %UPDATE_PACKED Fold one sample with values already in symbol_index order.
            obj.assertOpen();
            r = sentil_mex('monitor_update_packed', obj.Handle, double(time), double(values));
        end

        function reset(obj)
            %RESET Clear streaming state to run a fresh trace.
            obj.assertOpen();
            sentil_mex('monitor_reset', obj.Handle);
        end

        function f = formula(obj)
            %FORMULA A copy of the monitored formula.
            obj.assertOpen();
            f = sentil.Formula(sentil_mex('monitor_formula', obj.Handle));
        end

        function c = config(obj)
            %CONFIG A copy of the monitor's config.
            obj.assertOpen();
            c = sentil.Config(sentil_mex('monitor_config_of', obj.Handle));
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this monitor has been closed');
            end
        end
    end
end